building-and-packaging/material/cmake.md

```diff
@@ -0,0 +1,414 @@
+- Many GUIs and TUIs: ccmake, cmake-gui, integration in probably nearly all IDEs, ...
+- CMake is *"cross-platform"*: you can ideally use same cmake file in all OS's (easy to distinguish platform-specific things)
+- Build directory becomes independent of source directory (build multiple versions with different dependencies etc.)
+- CMake respects choices in user environment (e.g. user defines cpp compiler through CXX)
+
+### CMake "Hello World"
+
+Example adapted from [Yanson Tech](https://www.youtube.com/watch?v=wl2Srog-j7I)
+Go back to original `main.cpp`
+
+`CMakeLists.txt`:
+
+```cmake
+cmake_minimum_required(VERSION "3.12")
+
+project("HelloWorld")
+
+add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- CMake variable values: `${}`, use quotations to avoid problems with spaces
+- `mkdir build && cd build && cmake ..` ... explain files (TODO), look at makefile
+    - `CMakeCache.txt`
+- `make` and `./HelloWorld`
+
+### Multiple files
+
+- Add again `sse.cpp` and `sse.hpp` in subfolder `sse` and adapt `main.cpp`
+
+`CMakeLists.txt`:
+
+```diff
++ add_executable("${PROJECT_NAME}" main.cpp sse/sse.cpp)
+- add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- Generate and build again, this time both with verbose output TODO
+
+- What if there are multiple source files in `sse` folder?
+
+`CMakeLists.txt`:
+
+```diff
++ file(GLOB_RECURSE SRC_FILES CONFIGURE_DEPENDS main.cpp sse/*.cpp)
++ add_executable("${PROJECT_NAME}" "${SRC_FILES}")
+- add_executable("${PROJECT_NAME}" main.cpp sse.cpp)
+```
+
+- `GLOB_RECURSE`: filenames with wildcard characters, recurse -> also subdirectories
+- Actually discouraged since wildcard populated while generating. You need to run `cmake` again when you add a file. See [cmake docs](https://cmake.org/cmake/help/latest/command/file.html#filesystem).
+- `CONFIGURE_DEPENDS` partial remedy, but costly and no guarantee that all build systems support this
+
+### CTest
+
+- Let's add some dummy tests
+
+`main.cpp`:
+
+```diff
++  if (argc > 1)
++  {
++    return 1;
++  }
+```
+
+`CMakeLists.txt`:
+
+```diff
++ include(CTest)
++
++ add_test("test1" "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}")
++ add_test("test2" "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}" "someparam")
+```
+
+- `include` includes a module, some module are builtin in CMake
+- Creates target `test`
+- `make test`, explain output
+- We will learn later (after Christmas) how to work with proper packages for testing (e.g. boost)
+
+### CPack
+
+- We can also package code with cmake (or better: steer from cmake)
+
+`CMakeLists.txt`:
+
+```diff
+include(CPack)
+
+install(TARGETS "${PROJECT_NAME}" DESTINATION bin)
+```
+
+- Goes to `$CMAKE_INSTALL_PREFIX/bin`
+- Run cmake and `make package`, creates `HelloWorld-0.1.1-Linux.tar.gz`
+- We will see later (next week) how to package a `.deb`
+- Set version number
+
+```diff
++ SET(CPACK_PACKAGE_VERSION_MAJOR "1")
++ SET(CPACK_PACKAGE_VERSION_MINOR "2")
++ SET(CPACK_PACKAGE_VERSION_PATCH "3")
```

Comment:
Why does one have to do this manually? Will [CMake not derive these values](https://cmake.org/cmake/help/latest/module/CPack.html#variable:CPACK_PACKAGE_VERSION_MAJOR) from CMake variables/properties like `CMAKE_PROJECT_VERSION_MAJOR` or `PROJECT_VERSION_MAJOR`? I am not sure what the real variable names are.

Reply:
Good point :+1: